File: frontier/orchestrator/src/main/kotlin/dev/inprogress/frontier/Contracts.kt
package dev.inprogress.frontier

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.HexFormat
import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class ProbeInput(val value: String)

@Serializable
data class ProbeRequest(
  val operationId: String,
  val kind: String,
  val input: ProbeInput,
)

@Serializable
data class ProbeResult(
  val operationId: String,
  val value: String,
  val digest: String,
  val replayed: Boolean,
)

@Serializable
internal data class ExecutorErrorBody(val error: String)

internal object ProbeContract {
  const val KIND = "frontier-probe"
  const val MAX_VALUE_BYTES = 8 * 1024
  const val MAX_REQUEST_BYTES = 16 * 1024
  val json = Json { ignoreUnknownKeys = false }

  fun operationId(raw: String): String {
    val parsed = runCatching { UUID.fromString(raw) }.getOrNull()
    require(parsed != null && parsed.toString() == raw) { "workflow key must be a canonical UUID" }
    return raw
  }

  fun input(input: ProbeInput): ProbeInput {
    require(input.value.toByteArray(StandardCharsets.UTF_8).size <= MAX_VALUE_BYTES) {
      "probe value exceeds $MAX_VALUE_BYTES UTF-8 bytes"
    }
    return input
  }

  fun requestBody(request: ProbeRequest): String {
    val body = json.encodeToString(request)
    require(body.toByteArray(StandardCharsets.UTF_8).size <= MAX_REQUEST_BYTES) {
      "probe request exceeds $MAX_REQUEST_BYTES wire bytes"
    }
    return body
  }

  fun resultDigest(value: String): String =
    HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(value.toByteArray(StandardCharsets.UTF_8)))
}

File: frontier/orchestrator/src/contracts.ts
import { createHash } from "node:crypto";

import { z } from "zod";

export const PROBE_KIND = "frontier-probe" as const;
export const MAX_VALUE_BYTES = 8 * 1024;
export const MAX_REQUEST_BYTES = 16 * 1024;

const canonicalUuid = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

export const ProbeInputSchema = z
  .object({
    value: z.string().refine((value) => Buffer.byteLength(value, "utf8") <= MAX_VALUE_BYTES, {
      message: `probe value exceeds ${MAX_VALUE_BYTES} UTF-8 bytes`,
    }),
  })
  .strict();

export const ProbeRequestSchema = z
  .object({
    operationId: z.string().regex(canonicalUuid, "workflow key must be a canonical UUID"),
    kind: z.literal(PROBE_KIND),
    input: ProbeInputSchema,
  })
  .strict();

export const ProbeResultSchema = z
  .object({
    operationId: z.string().regex(canonicalUuid),
    value: z.string(),
    digest: z.string().regex(/^[0-9a-f]{64}$/),
    replayed: z.boolean(),
  })
  .strict();

export const ExecutorErrorSchema = z
  .object({
    error: z.enum(["invalid_request", "operation_id_conflict"]),
  })
  .strict();

export type ProbeInput = z.infer<typeof ProbeInputSchema>;
export type ProbeRequest = z.infer<typeof ProbeRequestSchema>;
export type ProbeResult = z.infer<typeof ProbeResultSchema>;

export function probeRequestBody(request: ProbeRequest): string {
  const body = JSON.stringify(ProbeRequestSchema.parse(request));
  if (Buffer.byteLength(body, "utf8") > MAX_REQUEST_BYTES) {
    throw new Error(`probe request exceeds ${MAX_REQUEST_BYTES} wire bytes`);
  }
  return body;
}

export function resultDigest(value: string): string {
  return createHash("sha256").update(value, "utf8").digest("hex");
}
